cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,287 @@
+            print(pad, f"Transformer-{1+log.transformer_id}: {log.transformer_name}", sep='')
+            print(pad, "Initial Circuit:", sep='')
+            print(textwrap.indent(str(log.initial_circuit), pad), "\n", sep='')
+            for log_level, log_text in log.logs:
+                if log_level.value >= level.value:
+                    print(pad, log_level, *log_text)
+            print("\n", pad, "Final Circuit:", sep='')
+            print(textwrap.indent(str(log.final_circuit), pad))
+            print("----------------------------------------")
+
+        done = [0] * self._curr_id
+        for i in range(self._curr_id):
+            # Iterative DFS.
+            stack = [(i, '')] if not done[i] else []
+            while len(stack) > 0:
+                log_id, pad = stack.pop()
+                print_log(self._logs[log_id], pad)
+                done[log_id] = True
+                for child_id in self._logs[log_id].nested_loggers[::-1]:
+                    stack.append((child_id, pad + ' ' * 4))
+
+
+class NoOpTransformerStatsLogger(TransformerStatsLogger):
+    """All calls to this logger are a no-op"""
+
+    def register_initial(self, circuit: 'cirq.AbstractCircuit', transformer_name: str) -> None:
+        pass
+
+    def log(self, *args: str, level: LogLevel = LogLevel.INFO) -> None:
+        pass
+
+    def register_final(self, circuit: 'cirq.AbstractCircuit', transformer_name: str) -> None:
+        pass
+
+    def show(self, level: LogLevel = LogLevel.INFO) -> None:
+        pass
+
+
+@dataclasses.dataclass()
+class TransformerContext:
+    """Stores common configurable options for transformers.
+
+    Args:
+        logger: `cirq.TransformerStatsLogger` instance, which is a stateful logger used for logging
+                the actions of individual transformer stages. The same logger instance should be
+                shared across different transformer calls.
+        ignore_tags: Tuple of tags which should be ignored while applying transformations on a
+                circuit. Transformers should not transform any operation marked with a tag that
+                belongs to this tuple. Note that any instance of a Hashable type (like `str`,
+                `cirq.VirtualTag` etc.) is a valid tag.
+    """
+
+    logger: TransformerStatsLogger = NoOpTransformerStatsLogger()
+    ignore_tags: Tuple[Hashable, ...] = ()
+
+
+TRANSFORMER = Callable[['cirq.AbstractCircuit', TransformerContext], 'cirq.Circuit']
+TRANSFORMER_TO_DECORATE = Callable[['cirq.Circuit', TransformerContext], 'cirq.Circuit']
+
+
+def _transform_and_log(
+    func: TRANSFORMER_TO_DECORATE,
+    transformer_name: str,
+    circuit: 'cirq.AbstractCircuit',
+    context: TransformerContext,
+) -> 'cirq.Circuit':
+    """Helper to log initial and final circuits before and after calling the transformer."""
+
+    context.logger.register_initial(circuit, transformer_name)
+    transformed_circuit = func(circuit.unfreeze(copy=True), context)
```

Comment:
The design doc notes that "The callable should not mutate the input circuit and instead return a copy." With the current implementation if I chain two transformer functions, the first has to allocate and return a new circuit, but then it gets copied here before being passed into the second transformer, even though the second transformer is not supposed to mutate it. This copy seems like unnecessary work.

It seems to me that transformer functions should accept and return `cirq.AbstractCircuit` instead. Since this abstract base type does not have any mutation methods, that makes clear in the signature that properly-implemented transformers can't mutate the circuit they receive, and we don't do any unnecessary copying or converting between formats. I know it's _possible_ that we might introduce new circuit implementations in the future, and some transformers may need to convert to a specific format to be implemented most efficiently, rather than using only what is available in the `AbstractCircuit` interface, but it still seems to me preferable to make any such copies or format conversions explicit in the transformers that need them, rather than trying to bake any default conversions or copies into the logging machinery, as here.

Reply:
My intention to do this change was that with the current proposed design, the callable's are free to mutate the input as we have already enforced "don't mutate the input" part in the decorator by passing a copy to the actual transformer.

This has the following benefits:
1.  The contract is very precise -- there's no scope of questions like "The method receives and returns an `AbstractCircuit`, will it preserve the type I pass in? Can I back `FrozenCircuit` if I pass a `Circuit`? etc.
2.  Returning an `AbstractCircuit` has an another disadvantage that we lose the circuit type information that was used to call the particular transformer. And hence, we'll have to introduce explicit casts at the call site, which will be annoying for users. For example:
```
@cirq.transformer
def custom_transformer(circuit: 'cirq.AbstractCircuit') -> 'cirq.AbstractCircuit':
    pass

def user_helper_method(circuit: 'cirq.Circuit') -> 'cirq.Circuit':
    # Do something. 
    return custom_transformer(circuit) # This breaks and need's an explicit cast.
```

3. The pattern followed today is that optimizers mutate the input. If we continue to allow this and silently add a copy in the decorator, it makes the migration of all user defined transformers to be just "add a @cirq.transformer decorator and your are good". 



Given the above benefits, my proposal would be to 
1. Keep `TRANSFORMER_TO_DECORATE` and `TRANSFORMER` as it is. 
2. Maybe add an explicit flag to the `@cirq.transformer(copy_input=True)` which can be used to control whether we copy the input by default or not. The default will be set to `True` and if we write a transformer where this is not needed, we can set the flag to `False` by default. 


WDYT ?